### PR TITLE
Add basic benchmarking infrastructure

### PR DESCRIPTION
**Pull request by tfiedor**
This PR adds basic infrastructure for profiling, performance testing and benchmarking:
  1. Series of testing and benchmarking automata.
  2. Templates and some basic programs to be compiled and profiled/performance tested.
  3. Benchmarking/perftesting scripts.

### CODE REVIEW

**Comment by tfiedor** [comment]
No need to rush with the reviews, if you have some ideas what other this could be there, then go for it.

In particular I'm wondering about:
  1. Should we include bigger benchmark (currently uncommited, about 10k of automata, 400MB)?
  2. Which functions we want to measure frequently (all? or some selected? like minimization, determinisation, binary operations, trims, etc.)

File: tests-performance/src/template-with-list-of-automata.cc
Comment (Adda0) [inline tests-performance/src/template-with-list-of-automata.cc]: Suggest using this template for your own benchmark automata.
```suggestion
     * Comment out automata, that you do not want to process or add your own automata.

```

File: tests-performance/src/template-with-list-of-automata.cc
Comment (Adda0) [inline tests-performance/src/template-with-list-of-automata.cc]: We should mention here that you are supposed to use `aut` and `alph` for your profiling operations. The same goes for all the templates which create `aut` for you.

File: tests-performance/src/complement.cc
Comment (Adda0) [inline tests-performance/src/complement.cc]: I would add a comment explaining that `comlement(aut, alph)` is the only statement that gets profiled.

File: tests-performance/src/template-with-cli-args.cc
Comment (Adda0) [inline tests-performance/src/template-with-cli-args.cc:1]: Could we do something about the indentation in these template files? It is all around the place, with 8 spaces alternating 4 spaces, if I see correctly. Was that supposed to be intentional? Cannot we separate the code chunks any other way? Using comments separating sections of the templates, for example?

File: tests-performance/pycobench
Comment (Adda0) [inline tests-performance/pycobench:1]: Do we want to just copy `pycobench` and `pyco_proc`? What if we want to update both to newer versions? Wouldn't a submodule pointing to their repository be more useful in the sense of maintainability?

File: tests-performance/src/template-with-cli-args.cc
Comment (Adda0) [inline tests-performance/src/template-with-cli-args.cc]: I would create a `constexpr std::string`.

File: tests-performance/src/template-with-cli-args.cc
Comment (Adda0) [inline tests-performance/src/template-with-cli-args.cc]: Then I can use `std::string::compare` instead of `strncmp` and `.size()` instead of `strlen` here. No need to fall back to C in this case.

File: tests-performance/README.md
Comment (Adda0) [inline tests-performance/README.md]: I would add a template which takes as a CLI parameter a file containing a list of paths to different automata to be tested. As with `single-automata.input`.

File: tests-performance/README.md
Comment (Adda0) [inline tests-performance/README.md]: Fix typos.
```suggestion
4. `pycobench`: a performance testing script
5. `pyco_proc`: a parser of performance results
```

**Review by Adda0 — COMMENTED** [review]
I am sure we will want to hear about the changes in our meeting. I will submit a review after the meeting and approve the PR when all additional changes discussed in the meeting are implemented as well.

**Comment by Adda0** [comment]
We will definitely want to perftest trim as it appears to be an expensive operation in Noodler. Moreover, we might want to think about perftesting overall performance of some of the basic automata operations such as adding new transitions, complement, determinisation, defragmentation, etc.

File: tests-performance/src/template-with-list-of-automata.cc
Comment (tfiedor) [reply tests-performance/src/template-with-list-of-automata.cc]: Good point, I will enhance the description. Or maybe add some example.

File: tests-performance/src/complement.cc
Comment (tfiedor) [reply tests-performance/src/complement.cc]: Sure.

File: tests-performance/src/template-with-list-of-automata.cc
Comment (Adda0) [reply tests-performance/src/template-with-list-of-automata.cc]: Both, I would say.

File: tests-performance/src/template-with-cli-args.cc
Comment (tfiedor) [reply tests-performance/src/template-with-cli-args.cc:1]: I will check it, didn't see any issues with indentation in my IDE, but there might be something wrong. I will try to unify it.

File: tests-performance/pycobench
Comment (tfiedor) [reply tests-performance/pycobench:1]: It will probably need some tweaks, so we should use the copy. We can discuss it during meeting.

File: tests-performance/src/template-with-cli-args.cc
Comment (tfiedor) [reply tests-performance/src/template-with-cli-args.cc]: Ook. Makes sense.

File: tests-performance/src/template-with-cli-args.cc
Comment (tfiedor) [reply tests-performance/src/template-with-cli-args.cc]: Ook. Makes sense.

File: tests-performance/README.md
Comment (tfiedor) [reply tests-performance/README.md]: Ook. Makes sense.

**Comment by tfiedor** [comment]
> We will definitely want to perftest trim as it appears to be an expensive operation in Noodler. Moreover, we might want to think about perftesting overall performance of some of the basic automata operations such as adding new transitions, complement, determinisation, defragmentation, etc.

Thanks, I will try to prepare some of sources/binaries for selected operation (binary operations, trims, etc.) we can then discuss some more complex scenarios at meeting.

**Comment by Adda0** [comment]
Sounds good to me. This way, we will have something to show. Additional operations can be added as we go, by anyone by simply copying the code used in these examples.

File: tests-performance/src/template-with-cli-args.cc
Comment (tfiedor) [reply tests-performance/src/template-with-cli-args.cc]: Sadly, in C++-17, it seems we cannot use `constexpr std::string` only `std::string`, so I will leave it as just `std::string`. Apparently `constexpr` strings are supported from C++20 only.

Comment (Adda0) [reply tests-performance/src/template-with-cli-args.cc]: Then we can use `const` at least. It does not matter much, but it will be nicer, in my opinion.

File: tests-performance/README.md
Comment (tfiedor) [reply tests-performance/README.md]: I decided to extend the template. If it is run without parameters it uses the hardcoded list of automata, otherwise it iterates all arguments each corresponding to one file, it opens the files and loads the automata in them.

File: tests-performance/README.md
Comment (Adda0) [inline tests-performance/README.md]: Fix typo
```suggestion
  * `template-with-list-of-automata.cc` contains basic template that without arguments takes a hardcoded list of input 

```

File: tests-performance/README.md
Comment (Adda0) [reply tests-performance/README.md]: It looks great. I consider this conversation to be resolved.

File: tests-performance/README.md
Comment (jurajsic) [inline tests-performance/README.md]: This should probably not be here

File: tests-performance/README.md
Comment (jurajsic) [inline tests-performance/README.md]: Why is there a line here?

**Review by jurajsic — APPROVED** [review]
It looks really good, I have just minor issues in README.

File: tests-performance/README.md
Comment (tfiedor) [reply tests-performance/README.md]: Why not? Can you elaborate?

Comment (jurajsic) [reply tests-performance/README.md]: It looks like it is copy-pasted from previous section and it does not fit there (there are two steps number 5, and the following part directly connects to number 4.

Comment (tfiedor) [reply tests-performance/README.md]: I reworded it.

Comment (Adda0) [reply tests-performance/README.md]: Have you really fixed it? I can still see the code block for `jobs.yaml` below step 5.

Comment (jurajsic) [reply tests-performance/README.md]: I think you changed something under `How to profile` which was all right, this stuff is under `How to test performance`.

Comment (tfiedor) [reply tests-performance/README.md]: Now really resolved. It is mentioned exhaustively in `scripts` part only.

**Comment by tfiedor** [comment]
@Adda0 I think, this is ready to be merged. You can check it one more time, and then merge it as you like.

I solved the multiple profiles of same configuration simply by counting how many similar files are there by `wc -l`, it is not foolproof, but the most simple IMO.

**Comment by Adda0** [comment]
> I solved the multiple profiles of same configuration simply by counting how many similar files are there by wc -l, it is not foolproof, but the most simple IMO.

Sounds good to me. Well done.

I will do a final review and merge today.

**Review by Adda0 — APPROVED** [review]
The PR looks great. Merging the PR now.

**Pull request merged**
